qiskit/circuit/quantumcircuit.py

```diff
@@ -1698,20 +1698,30 @@ def qasm(
                     operation = operation.copy(name=escaped)
 
                 # decompose gate using definitions if they are not defined in OpenQASM2
-                if (
-                    operation.name not in existing_gate_names
-                    and operation not in existing_composite_circuits
-                ):
-                    if operation.name in [
-                        operation.name for operation in existing_composite_circuits
-                    ]:
-                        # append operation id to name of operation copy to make it unique
-                        operation = operation.copy(name=f"{operation.name}_{id(operation)}")
-
-                    existing_composite_circuits.append(operation)
-                    _add_sub_instruction_to_existing_composite_circuits(
-                        operation, existing_gate_names, existing_composite_circuits
-                    )
+                if operation.name not in existing_gate_names:
+                    # If the operation has a unique qasm name, we set the name of the operation
+                    # to this qasm name. This is now the case for the permutation gate, for which
+                    # the unique qasm name is based on the permutation pattern.
+                    op_qasm_name = getattr(operation, "_qasm_name", None)
```

Comment:
Do we document this anywhere, like in the operation class or something? I think this might come as a surprise if I were to implement a new gate and wanted to serialize the qasm in a similar manner to permutation.

Reply:
Handling qasm for permutation gates is very heavily based on the code for unitary gates, except for the idea of assigning a name based on the pattern (which is already done for permutation circuits) without using an auxiliary id. What is your recommendation I should do here?

Reply:
Well it's more that this is an implicit API we're adding that if an `Operation` instance has a `_qasm_name` attribute that attribute is now used to specify the name of the gate when generating openqsm2. This is a new interface and we're not documenting it anywhere. I'm also a bit concerned about being tied to this moving forward as it kind of couples the operation object to its qasm representation which has been a source of problems with the `Instruction.qasm()` method (although this is much looser than the problems around `Instruction.qasm()`).

That being said I think maybe if we just leave this as a hack in place for `PermutationGate` for now that's enough so maybe something like:

```suggestion
                    if operation.name == "permutation":
                        op_qasm_name = getattr(operation, "_qasm_name")
```

(or maybe an `isinstance` check)

Reply:
Thanks for the explanation, and I agree, this is a cleaner "hack" that what I had. I also agree that we want a clean way (without hidden implicit assumptions) to specify qasm2 names for high-level-objects, which is somehow also in the spirit of other things we discussed: inverses, reversed_ops and so on.